apps/src/code-studio/components/progress/UnitCalendarDialog.jsx

```diff
@@ -0,0 +1,127 @@
+import React, {Component} from 'react';
+import PropTypes from 'prop-types';
+import i18n from '@cdo/locale';
+import BaseDialog from '@cdo/apps/templates/BaseDialog';
+import Button from '@cdo/apps/templates/Button';
+import UnitCalendar from './UnitCalendar';
+import _ from 'lodash';
+import {unitCalendarLesson} from '@cdo/apps/templates/progress/unitCalendarLessonShapes';
+
+const styles = {
+  dialog: {
+    textAlign: 'left',
+    paddingLeft: 20,
+    paddingRight: 20,
+    paddingBottom: 20
+  },
+  button: {
+    float: 'right',
+    marginTop: 30
+  }
+};
+
+export default class UnitCalendarDialog extends Component {
+  static propTypes = {
+    isOpen: PropTypes.bool,
+    handleClose: PropTypes.func,
+    lessons: PropTypes.arrayOf(unitCalendarLesson).isRequired,
+    weeklyInstructionalMinutes: PropTypes.number.isRequired
+  };
+
+  generateSchedule = () => {
+    const {lessons, weeklyInstructionalMinutes} = this.props;
+    const lessonsCopy = _.cloneDeep(lessons);
+    let allWeeks = [];
+    let currWeek = [];
+    let currMinutes = 0;
+
+    // splitting lessons across weeks
+    lessonsCopy.forEach(lesson => {
+      let lessonDuration = lesson.duration;
+      while (lessonDuration > 0) {
+        if (currMinutes + lessonDuration < weeklyInstructionalMinutes) {
+          // If the rest of the current lesson fits into this week, put it in the schedule.
+          currMinutes = currMinutes + lessonDuration;
+          currWeek.push(_.cloneDeep(lesson));
+          lessonDuration = 0;
+        } else if (currMinutes < weeklyInstructionalMinutes - 5) {
+          // If there's more than 5 minutes left in the week,
+          // add as much of the lesson as you can to this week.
+          let partialLesson = _.cloneDeep(lesson);
+          partialLesson.duration = weeklyInstructionalMinutes - currMinutes;
+          currWeek.push(partialLesson);
+          lessonDuration = lessonDuration - partialLesson.duration;
+          lesson.duration = lessonDuration;
+          currMinutes = weeklyInstructionalMinutes;
+        } else {
+          // If there isn't enough time in this week to add this lesson, start a new week.
+          allWeeks.push(currWeek);
+          currWeek = [];
+          currMinutes = 0;
+        }
+      }
+    });
+    allWeeks.push(currWeek);
+
+    // This is to determine where a lesson starts and ends and where the bulk of the lesson is.
```

Comment:
I'm wondering if we can do this as part of the step above so we don't have to loop through all the lesson chunks again. For example when you start looking at a new lesson you know that the first time you add that lesson isStart is true, when the duration is 0 then you no it is the end, and I think you could figure out majority by figuring out what the duration chunk was vs the remainder.  Also does this current implementation assume a lesson could not be split 3 times only 2?

Reply:
It can be split 3 times, we won't set a lesson as being the end until we see that a new lesson has started or we're out of lessons

Reply:
Oh ok great. If this is too crazy to combine no worries it just seemed like there might be a simpler way but it might not be worth trying to see if there is because this is working just fine.

Reply:
I found a way to combine it, let me know if you think this is cleaner!